File: util/uhdm-cmp.cpp
/*
 * Copyright 2019 Alain Dargelas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <uhdm/UhdmComparer.h>
#include <uhdm/uhdm-version.h>
#include <uhdm/uhdm.h>
#include <uhdm/vpi_visitor.h>

#include <algorithm>
#include <filesystem>
#include <iostream>
#include <memory>
#include <string>

namespace fs = std::filesystem;

static int32_t usage(const char *progName) {
  std::cerr << "Usage:" << std::endl
            << "  " << progName << " <uhdm-file> <uhdm-file>" << std::endl
            << std::endl
            << "Reads input uhdm binary representations of two files and "
               "compares them topographically. (Version: "
            << UHDM_VERSION_MAJOR << "." << UHDM_VERSION_MINOR << ") "
            << std::endl
            << std::endl
            << "Exits with code" << std::endl
            << "  = 0, if input files are equal" << std::endl
            << "  < 0, if input files are not equal" << std::endl
            << "  > 0, for any failures" << std::endl;
  return 0;
}

int32_t main(int32_t argc, char **argv) {
  if (argc != 3) {
    return usage(argv[0]);
  }

  fs::path fileA = argv[1];
  fs::path fileB = argv[2];

  std::error_code ec;
  if (!fs::is_regular_file(fileA, ec) || ec) {
    std::cerr << fileA << ": File does not exist!" << std::endl;
    return usage(argv[0]);
  }

  if (!fs::is_regular_file(fileB, ec) || ec) {
    std::cerr << fileB << ": File does not exist!" << std::endl;
    return usage(argv[0]);
  }

  std::unique_ptr<uhdm::Serializer> serializerA(new uhdm::Serializer);
  std::vector<vpiHandle> handlesA = serializerA->restore(fileA);

  std::unique_ptr<uhdm::Serializer> serializerB(new uhdm::Serializer);
  std::vector<vpiHandle> handlesB = serializerB->restore(fileB);

  if (handlesA.empty()) {
    std::cerr << fileA << ": Failed to load." << std::endl;
    return 1;
  }

  if (handlesB.empty()) {
    std::cerr << fileB << ": Failed to load." << std::endl;
    return 1;
  }

  if (handlesA.size() != handlesB.size()) {
    std::cerr << "Number of designs mismatch." << std::endl;
    return -1;
  }

  std::function<const uhdm::Design *(vpiHandle handle)> to_design =
      [](vpiHandle handle) {
        return (const uhdm::Design *)((const uhdm_handle *)handle)->object;
      };

  std::vector<const uhdm::Design *> designsA;
  designsA.reserve(handlesA.size());
  std::transform(handlesA.begin(), handlesA.end(), std::back_inserter(designsA),
                 to_design);

  std::vector<const uhdm::Design *> designsB;
  designsB.reserve(handlesB.size());
  std::transform(handlesB.begin(), handlesB.end(), std::back_inserter(designsB),
                 to_design);

  for (size_t i = 0, n = designsA.size(); i < n; ++i) {
    uhdm::UhdmComparer comparer;
    if (int32_t r = comparer.compare(designsA[i], designsB[i]); r != 0) {
      comparer.print(std::cout);
      return -1;
    }
  }

  return 0;
}

File: util/uhdm-cmp.py
from uhdm import uhdm
from uhdm import util

import argparse
import os
import sys


def main(uhdm_filea, uhdm_fileb):
    if not os.path.isfile(uhdm_filea):
        print(f"{uhdm_filea}: File does not exist!", file=sys.stderr)
        return 1
    sa = uhdm.Serializer()
    ha = sa.Restore(uhdm_filea)
    if not ha:
        print(f"{uhdm_filea}: Failed to load.")
        return 1

    if not os.path.isfile(uhdm_fileb):
        print(f"{uhdm_fileb}: File does not exist!", file=sys.stderr)
        return 1
    sb = uhdm.Serializer()
    hb = sb.Restore(uhdm_fileb)
    if not hb:
        print(f"{uhdm_fileb}: Failed to load.")
        return 1

    if len(ha) != len(hb):
        print(f"Number of designs mismatch.")
        return 1

    for i in range(len(hb)):
        context = uhdm.CompareContext()
        ca = uhdm.UhdmBaseClassFromVpiHandle(ha[i])
        cb = uhdm.UhdmBaseClassFromVpiHandle(hb[i])
        if ca.Compare(cb, context):
            if context.m_failedLhs:
                print(f"LHS: {uhdm.decompile(context.m_failedLhs)}")
            else:
                print(f"LHS: <null>")
            if context.m_failedRhs:
                print(f"RHS: {uhdm.decompile(context.m_failedRhs)}")
            else:
                print(f"RHS: <null>")
            return -1
        return 0


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description=(
            "Reads input UHDM binary representations of two files and "
            "compares them topographically. "
            f"(Version: {uhdm.UHDM_VERSION_MAJOR}.{uhdm.UHDM_VERSION_MINOR})\n\n"
            "Exits with code:\n"
            "  = 0, if input files are equal\n"
            "  < 0, if input files are not equal\n"
            "  > 0, for any failures"
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument("uhdm_filea", help="Path to the first UHDM file")
    parser.add_argument("uhdm_fileb", help="Path to the second UHDM file")

    args = parser.parse_args()
    uhdm_filea = args.uhdm_filea
    uhdm_fileb = args.uhdm_fileb

    rc = main(uhdm_filea, uhdm_fileb)
    sys.exit(rc)

